### PR TITLE
Buip055 txnsighash

### PR DESCRIPTION
**Pull request by gandrewstone**

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -1923,6 +1943,83 @@ bool CheckTxInputs(const CTransaction &tx, CValidationState &state, const CCoins
 }
 } // namespace Consensus
 
+bool CheckInputsAnalyzeTx(CTransaction &tx,
```
Comment (dgenr8) [inline src/main.cpp:1955]: Is there a way to do this with less repeated code? It may be hard to maintain.

File: src/primitives/transaction.h [hunk 2]
```diff
@@ -228,6 +228,8 @@ class CTransaction
     const std::vector<CTxOut> vout;
     const uint32_t nLockTime;
 
+    unsigned char sighashType; // BU memory only
```
Comment (dgenr8) [inline src/primitives/transaction.h:231]: Would another option be to add this to CTxMemPoolEntry instead of CTransaction?

File: qa/rpc-tests/test_framework/util.py [hunk 3]
```diff
@@ -150,28 +151,34 @@ def hex_str_to_bytes(hex_str):
...
     while True:
         pool = set(rpc_connections[0].getrawmempool())
         num_match = 1
+        poolLen = [len(pool)]
         for i in range(1, len(rpc_connections)):
-            if set(rpc_connections[i].getrawmempool()) == pool:
+            tmp = set(rpc_connections[i].getrawmempool())
+            if tmp == pool:
                 num_match = num_match+1
+            poolLen.append(len(tmp))
```
Comment (sickpig) [inline qa/rpc-tests/test_framework/util.py]: camelcase?

File: src/buip055fork.cpp [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+#include <atomic>
+#include <boost/foreach.hpp>
+#include <boost/lexical_cast.hpp>
+#include <boost/thread.hpp>
+#include <inttypes.h>
+#include <iomanip>
+#include <limits>
+#include <queue>
+
+std::vector<unsigned char> invalidOpReturn = {112, 101, 101, 114, 50, 112, 101, 101, 114, 32, 99, 97, 115, 104, 32, 114, 101, 113, 117, 105, 114, 101, 115, 32, 108, 97, 114, 103, 101, 114, 32, 98, 108, 111, 99, 107, 115};
```
Comment (sickpig) [inline src/buip055fork.cpp]: for clarity better use a const std::string rather than a vector of u char.  At least we should add a comment with the actual value of the string.

in any case the `invalidOpReturn` need to be changed per spec, the new string is:
Bitcoin: A Peer-to-Peer Electronic Cash System
see: https://github.com/Bitcoin-UAHF/spec/blob/master/uahf-technical-spec.md#req-6-1-disallow-special-op_return-marked-transactions-with-sunset-clause

File: src/buip055fork.cpp [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+#include "timedata.h"
+#include "tinyformat.h"
+#include "tweak.h"
+#include "txmempool.h"
+#include "ui_interface.h"
+#include "util.h"
+#include "utilmoneystr.h"
+#include "utilstrencodings.h"
+#include "validationinterface.h"
+#include "version.h"
```
Comment (sickpig) [inline src/buip055fork.cpp]: do we really need all this includes?

File: src/buip055fork.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            return state.DoS(100,
+                             error("transaction is invalid on BUIP055 chain"), REJECT_INVALID, "bad-txns-wrong-fork");
+    }
+    return true;
+}
+
+bool ValidateBUIP055Tx(const CTransaction& tx)
+{
+    bool ret = !IsTxOpReturnInvalid(tx);
+    // TODO: validate new signature (REQ-6-2, REQ-6-3)
```
Comment (sickpig) [inline src/buip055fork.cpp]: to be removed. implemented in `IsTxBUIP055Only`

File: src/buip055fork.cpp [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            }
+            if (pc != txout.scriptPubKey.end())
+            {
+                std::vector<unsigned char> data;
+                if (txout.scriptPubKey.GetOp(pc, op, data))
+                {
+                    // Note this code only works if the size <= 75 (or we'd have OP_PUSHDATAn instead)
+                    if (op == invalidOpReturn.size())
+                    {
+                        if (data == invalidOpReturn)
```
Comment (sickpig) [inline src/buip055fork.cpp]: TODO: add sunset clause `block height <= 530000`

File: src/rpc/protocol.h [hunk 8]
```diff
@@ -13,7 +13,7 @@
 #include <string>
 #include <boost/filesystem.hpp>
 
-#include <univalue.h>
+#include "univalue/include/univalue.h"
```
Comment (sickpig) [inline src/rpc/protocol.h:16]: weird

File: qa/rpc-tests/buip055.py [hunk 9]
```diff
@@ -0,0 +1,490 @@
...
+        self.testDefaults()
+        time.sleep(1)  # by controlling the forkTime, we know exactly what the median block will be
+        forkTime = self.testCli()  # also sets up parameters on nodes 0, 1 to to fork
+        time.sleep(2)  # Wait to ensure that every newly generated block is after the fork time.
+
+        base = [x.getblockcount() for x in self.nodes]
+        assert_equal(base, [base[0]] * 4)
+
+        # TEST REQ-3: that a <= 1 MB block is rejected by the fork nodes
+        # the rejection happens for the first block whose nTime is
```
Comment (ptschip) [inline qa/rpc-tests/buip055.py]: what is nTime?  looks like you were about to write it in  "block whose nTime is"  ?

File: src/parallel.cpp [hunk 10]
```diff
@@ -549,6 +549,10 @@ void HandleBlockMessageThread(CNode *pfrom, const string strCommand, const CBloc
             if (nDoS > 0)
                 dosMan.Misbehaving(pfrom, nDoS);
         }
+
+        // the current fork is bad due to this block so reset the best header to the best fully-validated block
+        // so we can download another fork of headers (and blocks).
+        pindexBestHeader = chainActive.Tip();
```
Comment (ptschip) [inline src/parallel.cpp]: is this making pindexBestHeader = chainactive  necessary?  Because we are now multithreaded, it is possible that while you were processing this block you received other headers already on a different chain (the good chain) and pindexBestHeaeder would already have more chainwork..now you're resetting it here to be less than what it should be ... Furthermore,   If the block fails validation, then pindexBestHeader will already be the chaintip, unless other headers have arrived...i don't think adding this is helps and actually adds a new problem.

File: src/main.cpp [hunk 11]
```diff
@@ -1451,7 +1458,7 @@ bool AcceptToMemoryPoolWorker(CTxMemPool &pool,
         // There is a similar check in CreateNewBlock() to prevent creating
         // invalid blocks, however allowing such transactions into the mempool
         // can be exploited as a DoS attack.
-        if (!CheckInputs(tx, state, view, true, MANDATORY_SCRIPT_VERIFY_FLAGS, true, NULL))
+        if (!CheckInputsAnalyzeTx(tx, state, view, true, MANDATORY_SCRIPT_VERIFY_FLAGS | forkVerifyFlags, true, NULL))
```
Comment (ptschip) [inline src/main.cpp:1464]: don't we have to add the !CheckInputsAnalyzeTx() check in ConnectBlock() also?

File: src/rpc/server.h [hunk 12]
```diff
@@ -18,7 +18,7 @@
 
 #include <boost/function.hpp>
 
-#include <univalue.h>
+#include "univalue/include/univalue.h"
```
Comment (sickpig) [inline src/rpc/server.h:22]: ditto

File: qa/rpc-tests/buip055.py [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+import test_framework.script as script
+import pdb
+import sys
+if sys.version_info[0] < 3:
+    raise "Use Python 3"
+import logging
+logging.basicConfig(format='%(asctime)s.%(levelname)s: %(message)s', level=logging.INFO)
+
+
+invalidOpReturn = hexlify(b'peer2peer cash requires larger blocks')
```
Comment (sickpig) [inline qa/rpc-tests/buip055.py]: need to be changed per spec, the new string is:
`Bitcoin: A Peer-to-Peer Electronic Cash System`
see: https://github.com/Bitcoin-UAHF/spec/blob/master/uahf-technical-spec.md#req-6-1-disallow-special-op_return-marked-transactions-with-sunset-clause

File: src/parallel.cpp (hunk 10 above)
Comment (ptschip) [reply src/parallel.cpp]: if you really need to do this here maybe you can try:

if (FindMostWorkChain()->nChainWork > chainActive.Tip()->nChainWork)
{
    pindexBestHeader = FindMostWorkChain();
}
else
{
    pindexBestHeader = chainActive.Tip();
}

**Comment by sickpig** [comment]
@gandrewstone is this PR meant to replace #697 and #698? if this is the case it would be better to close them

File: src/parallel.cpp (hunk 10 above)
Comment (sickpig) [reply src/parallel.cpp]: @ptschip is `FindMostWorkChain()` thread safe?

Comment (gandrewstone) [reply src/parallel.cpp]: @ptschip, take a look at main.cpp:SendMessages() lines 7380 to 7413, and in particular line 7400:
                // BU Bug fix for Core:  Don't start downloading headers unless our chain is shorter
                if (pindexStart->nHeight < pto->nStartingHeight)

This chunk of code only downloads from nodes that have longer headers than our pindexBestHeader.  But what if our pindexBestHeader is pointing to a blockchain minority fork that we have eliminated due to the "invalidateblock" RPC call or marked bad because of the BUIP055 forking logic?

In that case we will NEVER download from a node that is presenting the minority fork, because the (invalid) majority fork is longer.  So BU nodes get stuck at the fork point because they never ask for minority fork headers or blocks.

So the above code resets the pindexBestHeader to the best known-good block when it finds a fork situation (basically when it finds the abnormal situation where headers are good but blocks are bad).  Because we no longer know whether headers correspond to good blocks, have the code drop back to the best known good block's header.  Is there ever any situation where we want pindexBestHeader to point to the best header even if its blocks could be bad?

I thought no, so I made this more comprehensive fix.  But I could just as easily fix the "stuck at fork point bug" by not touching pindexBestHeader at all.  Instead I'd use chainActive.Tip()->nHeight on line 7400 rather than pindexBestHeader->nHeight.  This would change the semantics of the code to request headers from every node whose headers are later than our blocks all the time, rather than requesting from nodes whose headers are later than our headers...

What are your thoughts on this?

File: src/main.cpp (hunk 11 above)
Comment (gandrewstone) [reply src/main.cpp:1464]: The purpose of CheckInputsAnalyzeTx is to mark transactions going into the mempool as playable on one fork or the other, so that when we build (mine) a new block template we pick the right transactions.  So semantically CheckInputs is exactly the same as CheckInputsAnalyzeTx -- that is, there is no difference in whether the transaction is rejected or accepted between the two functions.

Transactions (targeted to just the fork) are accepted or rejected based on the "flags" parameter and whether it has SCRIPT_ENABLE_SIGHASH_FORKID enabled.

When ConnectBlock is run, we know the block and so we know whether the block is on the fork or not.  We can set the flag based on whether the block is on the fork, and so if the transaction is targeted to the wrong fork, its validation fails and so connectblock fails (its a bad block).
setting the flag happens at main.cpp:2558
    if (pindex->forkActivated(miningForkTime.value))
    {
        flags |= SCRIPT_ENABLE_SIGHASH_FORKID;
    }

But in the mempool case we don't know whether the next block will be the fork block or not.  We know that on this chain the next block will be the fork block, but a re-org could rewind the blockchain a bit and the nTime in the new blocks could mean that the next block is not the fork block.

So tl;dr; we only have to track whether the transaction is "fork only" if its being admitted to the mempool.  If its part of a block we can just invalidate the tx and block if its in the wrong place.

File: src/buip055fork.cpp (hunk 6 above)
Comment (gandrewstone) [reply src/buip055fork.cpp]: done

File: src/buip055fork.cpp (hunk 4 above)
Comment (gandrewstone) [reply src/buip055fork.cpp]: done

File: src/rpc/protocol.h (hunk 8 above)
Comment (gandrewstone) [reply src/rpc/protocol.h:16]: I was getting include errors, and when you think about it, univalue is not installed as a system include so why are they using <>?

File: qa/rpc-tests/buip055.py (hunk 13 above)
Comment (gandrewstone) [reply qa/rpc-tests/buip055.py]: fixed

File: src/main.cpp (hunk 1 above)
Comment (gandrewstone) [reply src/main.cpp:1955]: maybe, but this code can just be deleted in the next version (once we are beyond the fork)

File: src/parallel.cpp (hunk 10 above)
Comment (ptschip) [reply src/parallel.cpp]: i don't see anything wrong with what you did, setting pindexBestHeader to chainActive.Tip() is fine...but what's missing is that because we're multithreaded, while we were processing the block we could have received more headers both on the good and possibly the bad chain...The bad chain headers are not of interest because as soon as you invalidate the block those headers will not be considered when you do FindMostWorkChain()....however, if you received more headers on the good chain, now what you're doing is setting pindexBestHeader to be less than what it actually is...so that's why you need to do what is below and only set it to be equal to chainActive if no other headers have arrived:

if (FindMostWorkChain()->nChainWork > chainActive.Tip()->nChainWork)
{
pindexBestHeader = FindMostWorkChain();
}
else
{
pindexBestHeader = chainActive.Tip();
}
